### ISSUE TITLE
Attachments

### ISSUE THREAD

**Issue by turt2live** [turt2live/matrix-appservice-webhooks#17]
ref: https://api.slack.com/docs/message-attachments

The `fallback` should be treated as plain text (no formatting) and is used when the client (matrix) doesn't support the message. 

## Color / pretext
`color`: `good`, `warning`, `danger`, or any hex code like `#439FE0`

Slack renders this like: 
![image](https://user-images.githubusercontent.com/1190097/28078981-f9f2a97e-6623-11e7-872d-00a00f061e34.png)

we'll render it like this:
![image](https://user-images.githubusercontent.com/1190097/28079706-a566c090-6626-11e7-95d4-8a5ce2f239ee.png)
![image](https://user-images.githubusercontent.com/1190097/28079721-b12a9294-6626-11e7-913b-0936213956a9.png)

Attachments are otherwise rendered with a grey line. It's unfortunate that matrix doesn't support the line wrapping (pre-text) thing, but this will work for now.
```
This would be pretext<br>
<font color='#d9534f'>▌</font> danger is #d9534f<br>
<font color='#f0ad4e'>▌</font> warning is #f0ad4e<br>
<font color='#5cb85c'>▌</font> good is #5cb85c<br>
<font color='#439fe0'>▌</font> custom is #439fe0
```

Pretext is `pretext`

Use #f7f7f7 for default.

## Authors

`author_name`, `author_link`, `author_icon`

![image](https://user-images.githubusercontent.com/1190097/28081094-6248d1fe-662b-11e7-8b95-f3afb2aa3161.png)

`<img src='mxc://t2bot.io/hcSELkhLCNMRxLLTXKffPPSn' width='16' height='16'> <a href='https://matrix.to/#/user/@voyager:t2bot.io'>Matrix Traveler</a>`

Although the image may need cropping ahead of time.

# Titles

`title`, `title_link`
![image](https://user-images.githubusercontent.com/1190097/28081146-880c5c58-662b-11e7-9700-44b4f2c2678c.png)

Basically an h3 as an anchor

# Fields 

`fields` is an array of `{title, value, short}`. `short` is optional.
![image](https://user-images.githubusercontent.com/1190097/28081198-ad3fc6e0-662b-11e7-98a0-979d1e36f315.png)

# Images

`image_url`
![image](https://user-images.githubusercontent.com/1190097/28081212-bac6e8a2-662b-11e7-824c-53ede4900c6d.png)

`thumb_url`
![image](https://user-images.githubusercontent.com/1190097/28081221-c02bb44e-662b-11e7-893d-fb3294ded2b6.png)

# Footers

`footer` and `footer_icon` (text & image). Treat much like the authors
![image](https://user-images.githubusercontent.com/1190097/28081258-dbe50cd0-662b-11e7-9822-377ebf2d682c.png)

`ts` (timestamp) - seconds since epoch

**Comment by joenas** [comment]
Hey. There's a veeeery crude implementation [here](https://github.com/joenas/matrix-appservice-webhooks/tree/slack_attachments). Feel free to have a look and you know.. provide some feedback.

Some notes (apart from code being a mess)
- Images aren't working because they need to be uploaded
- The sidebar looks ok in the synapse webclient but not much so in Riot
- Fields doesn't really behave like Slack, they ignore the `short` thingy
- Maybe ditch the "bar" thing and use `blockquote`? Or just use the bar for title/pretext or something
- Maybe not use `h3` for title because margins and whatnot
- Stuff like `<https://honeybadger.io/path/to/event/|ReferenceError>` from Slacks Honeybadger example isn't working but that would be fixed in #14 I believe?

**Comment by joenas** [comment]
<img width="951" alt="screen shot 2017-09-13 at 16 15 56" src="https://user-images.githubusercontent.com/1104793/30382223-dc2720ca-989e-11e7-907a-26600e57dc19.png">

This is what it looks like in Riot atm

**Comment by turt2live** [comment]
The issue with using blockquotes would be that Riot strips all attempts at coloring it. Although that might be solved with a patch to Riot. I'll see if I can squeeze the patch in, but for now using blockquote should be fine.

**Comment by joenas** [comment]
That would be awesome! So ditch the `bar` thing?

Also I think I'll add a `<br>` after field title like in Slack. The table like formatting we'll have to live without tho

**Comment by turt2live** [comment]
Yea, ditch the bar thing and replace it with `<blockquote style="border-color: #whatever;">...</blockquote>`.

Technically Riot supports tables, they're just really ugly. A `<br>` is probably better.

**Comment by joenas** [comment]
<img width="895" alt="screen shot 2017-09-13 at 19 20 35" src="https://user-images.githubusercontent.com/1104793/30391157-bbb2bbd2-98b8-11e7-9e25-262c713b58f3.png">

Looks pretty decent?! 

About the images, I guess I could use the `util.uploadContentFromUrl` somehow?

**Comment by turt2live** [comment]
Yea, it's just a promise thing. The avatar code should give a somewhat dirty example on how to use it.

**Comment by joenas** [comment]
I think I might drop the `<h3>` though.. h4 looks like bold on iOS and slightly larger on Riot desktop

**Comment by joenas** [comment]
> Yea, it's just a promise thing. The avatar code should give a somewhat dirty example on how to use it.

Put it in its own layer, before slack_attachments? Is it possible for it to not resolve before the next layer starts doing its thing?
Also it needs `bridge`, which it guess we'd have to send as an argument to each layer?

**Comment by turt2live** [comment]
The layer can return a promise and it'll block on that.

**Comment by joenas** [comment]
<img width="859" alt="screen shot 2017-09-14 at 11 38 06" src="https://user-images.githubusercontent.com/1104793/30422775-384e3e04-9941-11e7-993d-8c0a7eaa37ba.png">


Image uploading done! Anything else or should I create a PR and you can tear my JS-skillz apart? 😛 
I opted to create a new class for the image uploading instead of a layer but injected in the layers.. kind of. It felt better to me but :man_shrugging:

**Comment by joenas** [comment]
Maybe squash the commits?

**Comment by turt2live** [comment]
Opening a PR would probably be best so it's easier to see the changes.

**Comment by turt2live** [comment]
Added. Thanks @joenas !

**Comment by werkkrew** [comment]
Is supporting the "thumb_url" or "icon_url" slack message parameters in the plans?  For example, I have a payload like this:
```

{
	"username":"PlexBot",
	"text":"Plex (Foo)\r\nAmerican Dharma was recently added to Plex.",
	"attachments":
	[{
		"title":"American Dharma (2019)",
		"text":"A portrait of controversial Breitbart honcho and Donald Trump advisor, Stephen K. Bannon.",
		"title_link":"https://www.imdb.com/title/tt8804284",
		"fields":
		[{
			"short":true,
			"value":"<https://www.imdb.com/title/tt8804284|IMDb>",
			"title":"View Details"
		}],
		"fallback":"Image for American Dharma (2019)",
		"thumb_url":"https://i.imgur.com/HzH2885.jpg"
	}],
	"icon_url":"https://i.imgur.com/iF8m9LE.png"
}
```
